tests/phpunit/tests/canonical/postStatus.php

```diff
@@ -0,0 +1,972 @@
+				'post_author'  => self::$users['content_author'],
+				'post_parent'  => self::$posts['trash']->ID,
+			)
+		);
+
+		// Page for trashing.
+		self::$posts['trash-page'] = $factory->post->create_and_get(
+			array(
+				'post_type'    => 'page',
+				'post_title'   => 'trash page',
+				'post_name'    => 'trash-page',
+				'post_status'  => 'publish',
+				'post_content' => "Prevent canonical redirect exposing page slugs.\n\n<!--nextpage-->Page 2",
+				'post_author'  => self::$users['content_author'],
+			)
+		);
+		wp_trash_post( self::$posts['trash']->ID );
+		wp_trash_post( self::$posts['trash-page']->ID );
+	}
+
+	function setUp() {
+		parent::setUp();
+		self::setup_custom_types();
+	}
+
+	/**
+	 * Set up a custom post type and private status.
+	 *
+	 * This needs to be called both in the class setup and
+	 * test setup.
+	 */
+	public static function setup_custom_types() {
+		// Register public custom post type.
+		register_post_type(
+			'a-public-cpt',
+			array(
+				'public'  => true,
+				'rewrite' => array(
+					'slug' => 'a-public-cpt',
+				),
+			)
+		);
+
+		// Register private custom post type.
+		register_post_type(
+			'a-private-cpt',
+			array(
+				'public'             => false,
+				'publicly_queryable' => false,
+				'rewrite'            => array(
+					'slug' => 'a-private-cpt',
+				),
+				'map_meta_cap'       => true,
+			)
+		);
+
+		// Register custom private post status.
+		register_post_status(
+			'a-private-status',
+			array(
+				'private' => true,
+			)
+		);
+	}
+
+	/**
+	 * Test canonical redirect does not reveal private posts presence.
+	 *
+	 * @ticket 5272
+	 * @dataProvider data_canonical_redirects_to_ugly_permalinks
+	 *
+	 * @param string $post_key  Post key used for creating fixtures.
+	 * @param string $user_role User role.
+	 * @param string $requested Requested URL.
+	 * @param string $expected  Expected URL.
+	 */
+	public function test_canonical_redirects_to_ugly_permalinks( $post_key, $user_role, $requested, $expected ) {
+		wp_set_current_user( self::$users[ $user_role ] );
+		$this->set_permalink_structure( '' );
+		$post = self::$posts[ $post_key ];
+		clean_post_cache( $post->ID );
+
+		/*
+		 * The dataProvider runs before the fixures are set up, therefore the
+		 * post object IDs are placeholders that needs to be replaced.
+		 */
+		$requested = str_replace( '%ID%', $post->ID, $requested );
+		$expected  = str_replace( '%ID%', $post->ID, $expected );
+
+		$this->assertCanonical( $requested, $expected );
+	}
+
+	/**
+	 * Data provider for test_canonical_redirects_to_ugly_permalinks.
+	 *
+	 * @return array[] Array of arguments for tests {
+	 *     @type string $post_key  Post key used for creating fixtures.
+	 *     @type string $user_role User role.
+	 *     @type string $requested Requested URL.
+	 *     @type string $expected  Expected URL.
+	 * }
+	 */
+	function data_canonical_redirects_to_ugly_permalinks() {
```

Comment:
Redo these dataproviders so they're readable. At the moment, very hard to tell what is going on.